File: python/linkedlist/MergeTwoSortedLists.py
'''
https://leetcode.com/problems/merge-two-sorted-lists/
You are given the heads of two sorted linked lists list1 and list2.
Merge the two lists into one sorted list. The list should be made by splicing together the nodes of the first two lists.
Return the head of the merged linked list.

Input: list1 = [1,2,4], list2 = [1,3,4]
Output: [1,1,2,3,4,4]
'''
from typing import Optional
from ListNode import ListNode
from LinkedListUtils import build_linked_list, print_linked_list

class MergeTwoSortedLists:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # Dummy node to simplify the head logic
        dummy = ListNode(-1)
        tail = dummy
        # Iterate while both lists have nodes
        while list1 and list2:
            if list1.val <= list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next
        # Attach the remaining part of the list that is not empty
        if list1:
            tail.next = list1
        if list2:
            tail.next = list2
            
        return dummy.next

if __name__ == "__main__":
    solution = MergeTwoSortedLists()
    
    # Test Case 1
    l1 = build_linked_list([1, 2, 4])
    l2 = build_linked_list([1, 3, 4])
    merged = solution.mergeTwoLists(l1, l2)
    print(f"Merged List: {print_linked_list(merged)}") # Expected: [1, 1, 2, 3, 4, 4]
    
    # Test Case 2: Empty lists
    l1 = build_linked_list([])
    l2 = build_linked_list([])
    merged = solution.mergeTwoLists(l1, l2)
    print(f"Merged Empty Lists: {print_linked_list(merged)}") # Expected: []
    
    # Test Case 3: One empty list
    l1 = build_linked_list([])
    l2 = build_linked_list([0])
    merged = solution.mergeTwoLists(l1, l2)
    print(f"Merged One Empty: {print_linked_list(merged)}") # Expected: [0]

File: src/main/java/com/naresh/a_dsalgo/af_linkedlist/problems/MergeTwoSortedLists.java
package com.naresh.a_dsalgo.af_linkedlist.problems;

/**
 * Problem: Merge Two Sorted Lists
 * Description: Merge two sorted linked lists and return it as a sorted list.
 */
public class MergeTwoSortedLists {


    /**
     * Algorithm: Use a dummy node and a tail pointer. Iterate through both lists, 
     * attaching the smaller node to the tail. After the loop, attach any remaining 
     * nodes from either list.
     */
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        // Pattern: Dummy Node / Iterative | Time: O(n + m), Space: O(1)
        var dummy = new ListNode(0);
        var merge = dummy;

        while (list1 != null && list2 != null) {
            if (list1.val <= list2.val) {
                merge.next = list1; // Attach list1 node
                list1 = list1.next;
            } else {
                merge.next = list2; // Attach list2 node
                list2 = list2.next;
            }
            merge = merge.next; // Advance merge
        }
        
        merge.next = (list1 != null) ? list1 : list2; // Attach remaining elements
        return dummy.next;
    }
    // FAANG Tip: Dummy nodes simplify edge cases (like empty lists). Mention that O(1) space is achieved by rearranging pointers rather than creating new nodes.

    public static void main(String[] args) {
        var sol = new MergeTwoSortedLists();
        
        // List 1: 1 -> 2 -> 4
        var l1 = new ListNode(1, new ListNode(2, new ListNode(4)));
        // List 2: 1 -> 3 -> 4
        var l2 = new ListNode(1, new ListNode(3, new ListNode(4)));
        print(l1);
        print(l2);
        var merged = sol.mergeTwoLists(l1, l2);
        print(merged); // Expected: 1 -> 1 -> 2 -> 3 -> 4 -> 4
    }

    private static void print(ListNode node) {
        while (node != null) {
            System.out.print(node.val + (node.next != null ? " -> " : ""));
            node = node.next;
        }
        System.out.println();
    }
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
}
